Dynacred 2 - Observables

ineiti opened the pull request:
This is the new dynacred implementation based on observables.
Instead of having to poll for changes in attributes, everything is now
based on new blocks arriving.
Every time a new block arrives, the modified instances will trigger
updates to all classes that depend on them.

This PR replaces the former dynacred library with a new library based
on observables.
The webapp has been updated to use it.
The mobile app has a copy of the old dynacred in it, as it hasn't been
updated to use the new library.

# Review:
As a rule of thumb, try to have 3h of review for this PR.

## Directories
As this is a big piece of code, I propose the following parts to be reviewed first
- dynacred/src
- dynacred/spec
- webapp/app
- @tharvik: webapp/src/app/api and webapp/cas (this shouldn't have changed)
- mobile - ignore that one, I did not try to make it work :(

## Manual UI testing
Then please verify the following works:
- run the OLD code locally (`make` in `webapp`), create a new device on demo.c4dt.org/omniledger, copy the URL, replace `https://demo.c4dt.org/omniledger` with `http://localhost:4200` and create an OLD identity
- run the NEW code locally (`make` in `webapp`), and visit localhost:4200 - make sure it imports correctly and that you can interact with the account
- log in to c4dt.org and matrix, but when the login screen comes, replace the `https://demo.c4dt.org/omniledger` with `localhost:4200` and verify you're able to log in
  * tharvik NB: you might want to get the latest version of the CAS config, with `curl https://demo.c4dt.org/omniledger/assets/cas.conf > webapp/src/assets/cas.conf`

tharvik commented:
> * run the OLD code locally (`make` in `webapp`), create a new device on demo.c4dt.org/omniledger, copy the URL, replace `https://demo.c4dt.org/omniledger` with `http://localhost:4200` and create an OLD identity

Not a new device but took an already existing one linked to `localhost:4200`

> * run the NEW code locally (`make` in `webapp`), and visit localhost:4200 - make sure it imports correctly and that you can interact with the account

Error with `while checking user: {Error}: Error: couldn't find correct key and/or credID at: main`

> * log in to c4dt.org and matrix, but when the login screen comes, replace the `https://demo.c4dt.org/omniledger` with `localhost:4200` and verify you're able to log in

Unable to continue

tharvik commented on dynacred/src/keypair.ts:
```diff
@@ -0,0 +1,48 @@
+}
+
+function isScalar(o: any): o is Scalar {
+    return "marshalBinary" in o;
```
That looks like anything coming from kyber, isn't there something more precise?

ineiti replied:
Changed it to `setBytes` and changed the signature of `isScalar`. 

What should be really done is to avoid uglifying `Scalar`, `Point`, `Buffer`, and others. Unfortunately I cannot find the reference in webpack anymore. All I can find is turning off _all_ mangling of class-names: https://github.com/webpack/webpack/issues/8132 - but I know I saw a post where it describes how to keep just _some_ classes complete.

ineiti replied:
Ah, found it: https://webpack.js.org/configuration/externals/ - this would mean that @dedis/cothority should put @dedis/kyber as externals, and @c4dt/dynacred should probably do the same for @dedis/cothority and @dedis/kyber. Then you can import @dedis/{cothority,kyber} in the webapp, and be sure that you're using the same version, and that the classes still keep their name.

tharvik commented on dynacred/src/user.ts:
```diff
@@ -0,0 +1,122 @@
+        return newKP;
+    }
+
+    // TODO: should this be here or in CredentialStructBS?
```
make sense in `User`, as it's not the `CredentialStructBS` that request it but the user wanted to create a new device.

ineiti replied:
OK, removed TODO

tharvik commented on webapp/src/app/explorer/explorer.component.ts:
```diff
@@ -1,46 +1,45 @@
-                    this.contact = await Contact.fromByzcoin(this.uData.bc, id);
-                    this.kind = "credential";
+                    this.credStruct = await this.bcs.retrieveCredentialStructBS(id);
+                    this.kind = "credentialObservable.ts";
```
that looks weird, it's the only `kind` changed and to a filename

ineiti replied:
I blame IntelliJ...

tharvik commented on webapp/src/lib/dialog-transaction.ts:
```diff
@@ -14,72 +23,59 @@ export interface IDialogTransactionConfig<T> {
-        setTimeout(() => this.startTransactions(), 200);
+    async ngAfterViewInit() {
+        // TODO: replace the setTimeout with ChangeDetectorRef
+        setTimeout(async () => {
```
why using a `setTimeout` and not awaiting or launching it directly?

ineiti replied:
there was an error when creating the dialog that had to do with changing the layout while it's still drawing it. But I cannot reproduce it anymore. So:
- `ngAfterViewInit` -> `ngOnInit`
- removed `setTimeout`

tharvik commented on webapp/src/app/byz-coin.service.ts:
```diff
@@ -0,0 +1,155 @@
+            this.user = await this.retrieveUserByDB();
+            return;
+        } catch (e) {
+            Log.warn("couldn't find dynacred2 user");
```
dropping error?

ineiti replied:
removed the catch here, as the callers to `loadUser` catch the error.

ineiti commented:
> Error with while checking user: {Error}: Error: couldn't find correct key and/or credID at: main

Hmm - I just tried with a 'new' device from an old account, migrating to a new account.

Can you try if this account is still a valid account when running it from master?

tharvik commented:
> > Hmm - I just tried with a 'new' device from an old account, migrating to a new account.
> 
> Can you try if this account is still a valid account when running it from master?

It is still a valid account (loading user & showing the admin view, as expected)

ineiti commented:
OK, the `while checking user` is in fact just a warning, and it should continue after that. Is there any other message after it?

ineiti commented on dynacred/src/builder.ts:
```diff
@@ -0,0 +1,274 @@
+        throw new Error("couldn't find correct key and/or credID at: " + dbBase);
+    }
+
+    async migrateUser(dbold: IStorage, dbBase = "main"): Promise<boolean> {
```
changed from returning bool to rejecting promise.

ineiti commented on dynacred/src/builder.ts:
```diff
@@ -0,0 +1,274 @@
+        return new DarcsBS(darcs);
+    }
+
+    async retrieveDarcBS(darcID: BehaviorSubject<InstanceID> | InstanceID):
```
Needs discussion: from my point of view, there can be three outcomes in the `retrieve.*` methods:
1. the object is found (happy-path)
2. the object doesn't exist (undefined)
3. some network error occurred (promise rejected)

Where would you fit in an `Error` class?

tharvik replied:
hum, so the second case is not an error, I though it was; nevermind then.

ineiti commented on dynacred/src/builder.ts:
```diff
@@ -0,0 +1,274 @@
+ *
+ * The use of `dbBase` is to allow for different users in the same IStorage.
+ */
+export class ByzCoinBuilder {
```
Renaming then. Proposals:
- `ByzCoin` - I prefer not, as this collides with `ByzCoinRPC`
- `Fetcher`
- `Retriever` - and rename all methods with `s/retrieve//`
Other proposals?

tharvik replied:
I like to have methods beginning with verbs, so not the latter; `Fetcher` is the last remaining

ineiti replied:
sold.

ineiti commented on dynacred/src/byzcoin/transaction.ts:
```diff
@@ -0,0 +1,63 @@
+
+    async send(signers: ISigner[][], wait = 0): Promise<[ClientTransaction, AddTxResponse]> {
+        const ctx = ClientTransaction.make(this.bc.getProtocolVersion(), ...this.instructions);
+        this.instructions = [];
```
Can you explain more? I don't see what should be changed / answered / commented here.

tharvik replied:
when sending a transaction and there is for example a network issue, one is unable to retry it as the instructions will be empty. that might be a strand behavior.

ineiti replied:
OK, now I see your point. The `Transaction` is in fact a `TransactionBuilder`, so it's normal it resets the instructions after `send`. But `send` could have the following signature:

```
send(signers: ISigner[][], wait = 0): Promise<[ClientTransaction, Promise<AddTxResponse>]>
```

This would mean:
- good: the caller always gets the `ClientTransaction` and can decide if she wants to re-submit it
- bad: the caller needs to `await` one of the returned objects:
```
    const [ctx, atxPr] = await builder.send(signers);
    const atx = await atxPr;
```

If you can remove the `bad` part from above, I'd be very happy ;)

ineiti commented on dynacred/src/calypso.ts:
```diff
@@ -0,0 +1,105 @@
+        await tx.sendCoins(10);
+
+        if (p) {
+            p(-75, "Collective Re-Encryption");
```
Added comment to `TProgress`:
```
// If percentage is positive, the 'text' represents a transaction that is shown to move to the nodes, and then
// in the block.
// If percentage is negative, the 'text' represents a query and is only shown in the progress bar.
```

ineiti commented on dynacred/src/credentialStructBS.ts:
```diff
@@ -0,0 +1,508 @@
+export class CredentialStructBS extends BehaviorSubject<CredentialStruct> {
+    static readonly structVersionLatest = 2;
+    static readonly urlRegistered = "https://pop.dedis.ch/qrcode/identity-2";
+    static readonly urlUnregistered = "https://pop.dedis.ch/qrcode/unregistered-2";
```
replaced it with a URL pointing to a non-existing subdomain. Makes it eventually possible to add a service to show these QRCodes if opened in a browser.

ineiti commented on dynacred/src/credentialStructBS.ts:
```diff
@@ -0,0 +1,508 @@
+            try {
+                return PointFactory.fromProto(buf);
+            } catch (e) {
+                return undefined;
```
No, because the `getAttributeBS` returns a `Buffer.alloc(0)` if the attribute is not found. Which is the 'empty' attribute - except for the `Point`, where there is no such thing, as we don't even know which group we should use.

Changing `AttributePointBS` to return `Point|undefined`.

ineiti commented on dynacred/package.json:
```diff
@@ -1,66 +1,47 @@
   ],
-  "author": "C4DT",
-  "license": "GPL-3.0-or-later",
+  "author": "Linus Gasser",
```
it's in the github.com/c4dt organisation, but the main author (currently) is me...

ineiti commented on dynacred/src/builder.ts:
```diff
@@ -0,0 +1,274 @@
+            flatMap((id) => this.bc.instanceObservable(id)),
+            catchError((err) => {
+                Log.error(err);
+                return Promise.resolve(new BehaviorSubject(
```
I'm still trying to come up with an explanation why I did this... I think it's to cover an error in `bc.instanceObservable`, where a non-matching proof can be stored in the cache, but here we hope that the instance turned up in the meantime, and will be added with the next update from the chain...

tharvik replied:
isn't there a `filter` or smth to handle this kind of cases?

ineiti replied:
how would you use a filter here? The `throw new Error` is in the `flatMap` above, so it needs to be caught. And rxjs catches error using `catchError`...

Any suggestion welcome ;)

tharvik replied:
all good then

tharvik commented:
> OK, the `while checking user` is in fact just a warning, and it should continue after that. Is there any other message after it?

here goes the full log

```
Angular is running in the development mode. Call enableProdMode() to enable the production mode. core.js:25638
Firefox can’t establish a connection to the server at wss://fairywren.ch:7771/Status/Request. websocket-adapter.ts:61:18
 2: vendor.js:367621:23 [angular]  ->     Loaded latest block from db: {number}: 54243 log.ts:19:16
 2: vendor.js:367621:23 [angular]  ->     storing latest block in db: {number}: 54243 log.ts:19:16
 2: vendor.js:367621:23 [angular]  ->     Starting to update blocks for viewer log.ts:19:16
W : vendor.js:367621:23 [angular]  -> while checking user: {Error}: Error: couldn't find correct key and/or credID at: main log.ts:19:16
 2: main.js:894:30 [angular]       ->     Starting new subscription for newBlocks log.ts:19:16
Firefox can’t establish a connection to the server at wss://fairywren.ch:7771/Status/Request. websocket-adapter.ts:61:18
Firefox can’t establish a connection to the server at wss://conode.c4dt.org:7771/Status/Request. websocket-adapter.ts:61:18
```

tharvik commented:
After updating the user data (such as changing the email address), it added a `version` field to the struct in StorageDB, which make it works.

ineiti commented:
@tharvik - did a first round over your comments. If you're happy, `resolve` them. Else add to the discussion.

I modified the migration to also work in the absence of `version` and added a test to make sure it works.

cgrigis commented:
> Then please verify the following works:
> 
>     * run the OLD code locally (`make` in `webapp`), create a new device on demo.c4dt.org/omniledger, copy the URL, replace `https://demo.c4dt.org/omniledger` with `http://localhost:4200` and create an OLD identity

By "OLD code", I presume the `master` branch?
With a fresh clone of `omniledger`, `cd webapp`, `make`, when I paste the registration url with localhost replaced I get this in the console:

```
 Error: "Uncaught (in promise): fetching config gave: 404: [object ReadableStream]"
```

Do I need a `config.toml` somewhere?

ineiti commented:
Oups - yes. I added a config.dedis.toml in the latest commit. Copy it to `config.toml` in `omniledger/webapp/src/assets`.

ineiti commented on dynacred/DYNACRED.md:
```diff
@@ -0,0 +1,228 @@
+*   that defines the rules to execute the available actions has data that is stored in all nodes User accounts are a set
+*   of different instances linked together as described in the rest of this document. One difference from other
+*   blockchains is the possibility to easily use different devices with separated keys that can be revoked on a
+*   per-device basis.
```
bad copy/paste from google-doc...

ineiti commented on dynacred/DYNACRED.md:
```diff
@@ -0,0 +1,228 @@
+*   of all DARCs that are allowed to recover the account Calypso: a list of all stored secrets of this user Coin - every
+*   operation on OmniLedger has a cost that will be deduced from the coin account of the user. Currently only the
+*   creator of the OmniLedger network can fill up the coin accounts CalypsoWrite - a secret stored in OmniLedger with
+*   the possibility of delegating decryption access
```
fixed

ineiti commented on dynacred/DYNACRED.md:
```diff
@@ -0,0 +1,228 @@
+*   own darc that defines the rules specific to that instance. This is mostly for easier handling in the backend and to
+*   separate the devices-darcs from the other darcs. Each device darc represents one device that the user holds. A user
+*   can give access to his structure by adding a new device-darc with a new public key, and changing the __sign_ and
+*   _invoke:darc.evolve_ rules of the signer darc The _signer_ darc is referenced by all the other darcs in the user.
```
fixed

ineiti commented on dynacred/README.md:
````diff
@@ -0,0 +1,320 @@
+Here a small overview how the different instances are used on OmniLedger:
+
+```ascii
+
````
replaced by DYNACRED.md in fact. Removed that paragraph.

ineiti commented on dynacred/README.md:
```diff
@@ -0,0 +1,320 @@
+To simplify the collection and the signature of these `Instruction`s, dynacred uses the `CredentialTransaction` class.
+It abstracts the collection of the instructions and the signing of the final `ClientTransaction`.
+
+To modify a structure in dynacred, the following steps happen:
+1. client: create a `CredentialTransaction`: `const tx = user.startTransaction()`
+2. client: invoke the method of the class: `user.coinBS.transferCoins(tx, destID, Long.from(1000))`
+3. client: finalize the transaction and send it to ByzCoin: `tx.sendCoins()`
+4. byzcoin: verify the transaction and put it in a block
+5. byzcoin: update the global state
+6. byzcoin: inform the client a new block is available
+7. client (background): send a list of all instances and the known version to byzcoin
+8. byzcoin: reply with the instances that changed
+9. client (background): update all changed instances and inform the `Observable` of the `InstanceBS`
+10. client (background): update the classes and inform the `Observable` of `coinBS`
+11. client (background): if the `coinBS` is included in angular using `(coinBS | async).value`, the value will be 
+automatically updated
```
good idea - I think we don't have a good way yet to create figures / drawings from text... Any suggestions? Else I'll just do a figure in google docs, but that's not editable...

cgrigis replied:
Yes we do :)  [PlantUML](https://plantuml.com/) can create nice UML diagrams from text (see https://github.com/c4dt/service-stainless/blob/master/seq-diagram.txt for an example).

ineiti replied:
Added figure to README.

ineiti commented:
> I like fetcher.
> But as I understand it, the behaviour is not only to fetch, but it can also migrate...
> As it seems to be somehow an abstraction for the ByzCoin storage, ByzCoinDB? ByzCoinData?

`BCFetcher`? But then we lose the `fetch` verb at the beginning of the methods.
I can also move the `migrate` elsewhere, e.g., as a static method to `User`.

ineiti commented on dynacred/src/builder.ts:
```diff
@@ -0,0 +1,274 @@
+        const credID = Buffer.from(url.query.credentialIID, "hex");
+        const ephemeralBuf = Buffer.from(url.query.ephemeral, "hex");
+        if (credID.length !== 32 || ephemeralBuf.length !== 32) {
+            throw new Error("either credentialIID or ephemeral is not of length 32 bytes");
```
Neither is supposed to happen...

ineiti commented on dynacred/src/builder.ts:
```diff
@@ -0,0 +1,274 @@
+    async retrieveDarcBS(darcID: BehaviorSubject<InstanceID> | InstanceID):
+        Promise<DarcBS | undefined> {
+        Log.lvl3("getting darcBS");
+        if (darcID instanceof Buffer) {
```
yes? Why not? if it's a `BehaviorSubject`, then it can change over time (which is probably rare, in fact). If it's an `InstanceID` (which is a `Buffer`), then it's fixed.

cgrigis replied:
> `InstanceID` (which is a `Buffer`)

Aha, that was the part I was missing.
Shouldn't it then be `if ... instanceof InstanceID` for clarity?

cgrigis replied:
Furthermore (but this is personal preference), instead of calling a function with various possible different types and hoping the function does the right thing, I find it clearer to either have a polymorphic type, or do:
```ts
async retrieveDarcBSwithID(darcID: InstanceID): Promise<DarcBS | undefined> {
    return await retrieveDarcBS(new BehaviorSubject(darcID));
}

async retrieveDarcBS(darcID: BehaviorSubject<InstanceID>): Promise<DarcBS | undefined> {
    ...
}
```

ineiti replied:
Yes, using two methods is the go way. But I like polymorphism ;)

I tried `darcID instanceof InstanceID`, but it doesn't work. So I added a comment:
```
        // Need to verify against Buffer here, which is the defined type of InstanceID.
        // Else typescript complains....
```

cgrigis replied:
> Yes, using two methods is the go way. But I like polymorphism ;)

But in an OO language, the polymorphism usually employs a class hierarchy, where a different subtype implements a common method in their appropriate way: this removes the need to explicitly check for the type in the using code, and is transparently extended if new subtypes appear.

> I tried `darcID instanceof InstanceID`, but it doesn't work.

That's another bad point for TypeScript ;-)  Just out of curiosity, what does the error say?

> So I added a comment:
> 
> ```
>         // Need to verify against Buffer here, which is the defined type of InstanceID.
>         // Else typescript complains....
> ```

Thank you, better than nothing. :)

ineiti replied:
```
TS2693: 'InstanceID' only refers to a type, but is being used as a value here. 
```

And, yes, that's not polymorphism.

ineiti commented on dynacred/src/byzcoin/transaction.ts:
```diff
@@ -0,0 +1,63 @@
+        return inst;
+    }
+
+    unshift(inst: Instruction): Instruction {
```
My brain is javascript-enabled ;) - changed. Not my brain, the code...

cgrigis replied:
> My brain is javascript-enabled ;)

That's scary! ;-)

ineiti commented on dynacred/src/calypso.ts:
```diff
@@ -0,0 +1,105 @@
+import { CredentialTransaction } from "./credentialTransaction";
+import { KeyPair } from "./keypair";
+
+// Copied from webapp/lib/UI.ts - should probably be merged in CredentialTransaction and then be used
+// in showTransactions - which should also be revamped. But that is for later...
```
Will have to include it somehow in this code - probably will go into the newly named `SpawnerTransactionBuilder`...

ineiti replied:
removed it for the moment - else it gets too complicated.

ineiti commented on dynacred/src/calypso.ts:
```diff
@@ -0,0 +1,105 @@
+            p(50, "Decryption Request");
+        }
+        const rdID = tx.spawnCalypsoRead(wrID, kp.pub);
+        await tx.sendCoins(10);
```
yes - even though 7 would be more magic ;)

cgrigis replied:
Not if aliens with 7 fingers read this code ;-)

ineiti commented:
@cgrigis - same as for Valérian: if you're happy with the changes, `Resolve conversation`.
